pkg/mutation/system_external_data.go

```diff
@@ -0,0 +1,242 @@
+	var (
+		wg    sync.WaitGroup
+		mutex sync.RWMutex
+		fn    = s.sendRequestToExternalDataProvider
+
+		// the provider name is the first key and the outbound data is the second key
+		responses = make(map[string]map[string]*externaldata.Item)
+		// errors that might have occurred per provider
+		errors = make(map[string]error)
+	)
+
+	if fn == nil {
+		fn = defaultSendRequestToExternalDataProvider
+	}
+
+	for name, keys := range providerKeys {
+		provider, err := s.providerCache.Get(name)
+		if err != nil {
+			log.Error(err, "failed to get external data provider", "provider", name)
+			continue
+		}
+
+		wg.Add(1)
+		go func(provider *v1alpha1.Provider, keys []string) {
+			defer wg.Done()
+
+			resp, err := fn(context.Background(), provider, keys)
+
+			mutex.Lock()
+			defer mutex.Unlock()
+
+			if err != nil {
+				errors[provider.Name] = fmt.Errorf("failed to send external data request to provider %s: %w", provider.Name, err)
+				return
+			}
+			if err := validateExternalDataResponse(resp); err != nil {
+				errors[provider.Name] = fmt.Errorf("failed to validate external data response from provider %s: %w", provider.Name, err)
+				return
+			}
+
+			responses[provider.Name] = make(map[string]*externaldata.Item)
+			for _, item := range resp.Response.Items {
+				item := item // for scoping
+				responses[provider.Name][item.Key] = &item
+			}
+		}(&provider, keys.List())
+		wg.Wait()
+	}
+
+	return responses, errors
+}
+
+// mutateWithExternalData recursively traverses the given object and replaces
+// all external data placeholders with the corresponding external data items.
+func (s *System) mutateWithExternalData(object *unstructured.Unstructured, externalData map[string]map[string]*externaldata.Item, errors map[string]error) error {
+	var mutate func(interface{}) []error
+	mutate = func(current interface{}) []error {
+		var allErrors []error
+		switch obj := current.(type) {
+		case map[string]interface{}:
+			for k, v := range obj {
+				placeholder, ok := v.(*unversioned.ExternalDataPlaceholder)
+				if !ok {
+					// not a placeholder, let's continue recursing
+					allErrors = append(allErrors, mutate(v)...)
+					continue
+				}
+
+				// our base case - we found a placeholder and we should resolve it
+				var data *externaldata.Item
+				var providerResponse map[string]*externaldata.Item
+
+				err := errors[placeholder.Ref.Provider]
+				if err == nil {
+					providerResponse, ok = externalData[placeholder.Ref.Provider]
+					if !ok {
+						err = fmt.Errorf("failed to find external data provider %s in responses", placeholder.Ref.Provider)
+					}
+				}
+
+				if providerResponse != nil {
+					// we expect the response to contain the key we're looking for
+					data, ok = providerResponse[placeholder.ValueAtLocation]
+					if !ok {
+						err = fmt.Errorf("failed to find external data item in response for provider %s", placeholder.Ref.Provider)
```

Comment:
will this clobber the error taken from the `errors` map?

Reply:
It shouldn't clobber. There are four places we can potentially assign a value to `err` in order:

1. error found in the `errors` map
2. unable to find the provider name in `externalData` map
3. unable to find `placeholder.ValueAtLocation` in `providerResponse` map
4. error found within the data / unable to convert data value to string

Since each step depends on its previous steps, we would have terminated early if err is not nil. I can add `err == nil` as part of all if statement conditions for sanity.